lib/easybill_rest_client/request.rb

```diff
@@ -0,0 +1,96 @@
+# frozen_string_literal: true
+require 'easybill_rest_client/request_logger'
+require 'easybill_rest_client/retry_on'
+
+module EasybillRestClient
+  class Request
+    BASE_URL = 'https://api.easybill.de/rest/v1'
+    USERNAME = 'rest-api@easybill.de'
+    SUPPORTED_METHODS = %i(get post put delete).freeze
+    OPEN_TIMEOUT = 5
+
+    def initialize(opts = {})
+      unless SUPPORTED_METHODS.include?(opts.fetch(:method).to_sym)
+        raise ArgumentError, "Unsupported HTTP method '#{method}'"
+      end
+
+      @api_key = opts.fetch(:api_key)
+      @method = opts.fetch(:method)
+      @endpoint = opts.fetch(:endpoint)
+      @params = opts.fetch(:params)
+      @logger = opts.fetch(:logger)
+      @tries = opts.fetch(:tries)
+      @retry_cool_off_time = opts.fetch(:retry_cool_off_time)
+    end
+
+    def run
+      request_logger.info("#{method.to_s.upcase} #{endpoint} #{request_details}")
+      retry_on(TooManyRequests) do
+        retry_on(Net::OpenTimeout) do
+          response = perform_request
+          raise TooManyRequests if response.is_a?(Net::HTTPTooManyRequests)
+          response
+        end
+      end
+    end
+
+    def request_details
+      body_allowed? ? request.body : uri.query
+    end
+
+    private
+
+    attr_reader :api_key, :method, :endpoint, :params, :logger, :request_logger, :retry_on, :tries,
+      :retry_cool_off_time
+
+    def perform_request
+      Net::HTTP.start(uri.host, uri.port, use_ssl: uri.scheme == 'https') do |http|
+        http.open_timeout = OPEN_TIMEOUT
```

Comment:
FYI, you can provide it as an option to `.start`, i.e. `Net::HTTP.start(..., use_ssl: ..., open_timeout: OPEN_TIMEOUT)`.
Not sure even if your way is correct, because I think connection is already open inside the block, so the timeout may be ignored.